오픈 마켓 [STEP 3] 릴리, 예거

yeahg-dev opened the pull request:
> 오픈마켓이 1, 2로 나뉘어졌으나 브랜치가 달리지지 않고, 같은 팀을 유지해서 진행하기 때문에 리뷰어와 합의 후 STEP3, STEP4 로 지칭하여 PR을 제출하기로 했습니다. 참고바랍니다~

안녕하세요 찰리~~!💜🐻💜 @kcharliek 
바쁘심에도 불구하고 중간 중간 진행상황체크 해주시고, 조언주셔서 감사합니다🥰
아직 미구현한 기능이 좀 있지만, 구현한 부분까지 정리하여 스텝3 제출합니다.
이번 스텝도 잘 부탁드립니다🙏🏻
<br>

# 📱 `스텝3` 구현 내용

|- imagePicker 기능 -|- refreshControl 기능 -|- Keyboard -|
|:-:|:-:|:-:|
|<img src="https://user-images.githubusercontent.com/71127966/150973139-26a49552-ba3d-4639-a7aa-7adb492a57e4.gif">|<img src="https://user-images.githubusercontent.com/71127966/150974350-7c134d69-c927-4a15-96fe-f5147f08a943.gif">|<img src="https://user-images.githubusercontent.com/71127966/150974083-f2bdbcc4-aa3b-42a9-81f7-b5a0ae9f4b81.gif">|
|imagePicker Cell 을 클릭하면<br>사진첩에서 사진을 추가할 수 있습니다.<br>정방형으로 크롭이 가능합니다.<br>사진은 최대 5장까지 추가할 수 있습니다.|화면의 edge 를 당겨서(pulling)<br>새로고침(refresh) 할 수 있습니다.<br>상품등록에 성공하면 Modal 이 내려가고<br>화면이 새로고침되며 추가한 상품을 보여줍니다.|사용자가 textField/textView 를 터치하면<br>입력하는 데이터에 알맞은 키보드가 올라오고<br>키보드의 높이만큼 화면도 함께 올라갑니다.|

<br>

# ⚔️ `스텝3` 구현 방향성 논의

## 1️⃣ 상품 등록/수정 UI 를 하나의 뷰컨 안에서 분기처리 🆚 각각 다른 뷰컨 사용해서 구현(선택 ✅) 

- 지금은 두 화면이 유사하지만, 만약 현업에서 개발자로 일한다면 앞으로 화면에 기능이나 디자인이 추가되면서 계속 `분기처리`가 늘어날 리스크가 있으므로, 미리 뷰컨을 나눠두는 게 좋을 거라 생각했습니다.
- 객체지향 원칙 중 `단일책임원칙`에 의거해, 뷰컨을 미리 분리하는 게 맞다고 판단했습니다.
- 🥲 아쉬운 점은, 이번 스텝3 PR 에 상품 수정 UI 는 포함되어 있지 않습니다. 시간관계상, `상품 등록` 까지만 성공해서 제출하는 점 참고 부탁드립니다!

## 2️⃣ 상품 등록 화면을 Full Screen Modal 로 보여주기(선택 ✅) 🆚 각 뷰컨에 Navigation Controller 설치

- <오픈 마켓> 프로젝트는 등록된 상품을 테이블뷰와 컬렉션뷰 형태로 보는 것이 큰 맥락이라고 볼 수 있습니다.
- `상품 등록`은 큰 맥락이라기 보다, 잠깐 샛길로 빠져서 등록하고 돌아오는 형태를 취하므로, `Present Modally`가 사용자 경험 측면에서 적합하다고 생각했습니다.
- 이때, Modal 에는 Navigation Controller 가 적용되지 않으므로, 제목을 표시할 `Navigation Bar`와 [Cancel], [Done] 버튼을 표시할 `Bar Button Item`은 스토리보드의 라이브러리에서 꺼내 Modal 화면에 직접 붙여줬습니다.

## 3️⃣ 상품 등록 화면 전체를 테이블뷰로 구현 🆚 스크롤뷰 안에 스택뷰로 구현(선택 ✅)

- 테이블뷰나 컬렉션뷰를 사용하는 이유는 `셀의 재사용` 때문입니다. 상품 등록 화면은 셀을 재사용해야 할 만큼 내용이 많지 않습니다.
- 또한 재사용을 하기엔 UI 컴포넌트 생김새가 겹친다고 보기 어려웠습니다.
- 난이도 면에서, 테이블뷰보다는 `스크롤뷰 + 스택뷰`가 나아보였습니다.

## 4️⃣ UIImagePickerController 부분을 가로 스크롤이 가능한 컬렉션뷰로 구현(선택 ✅) 🆚 스택뷰

- 컬렉션뷰를 이용하면, 사진 추가를 위한 버튼에 `didSelectItemAt` 메서드를 활용할 수 있으므로, 별도의 탭제스처가 필요하지 않습니다.
- 이미지를 추가하고 `reloadData` 메서드를 호출해, 추가한 사진을 바로 화면에 보여줄 수 있습니다.
<br>

# 🤔고민한 부분

## 1️⃣ 상품 등록 성공시, 상품 리스트 뷰컨트롤러를 최신 상품리스트로 업데이트하는 방법
`상품 등록 화면(ProductRegisterationViewController)`에서 `상품 리스트 화면(ProductTableViewController, ProductCollectionViewController)`으로 전환했을 때,
**항상 서버의 최신 상품리스트 정보를 받아와서 상품 리스트를 보여주고 싶었습니다.**

크게 고민되었던 포인트는 두가지가 있었습니다.
- 상품리스트 뷰컨트롤러가 어떻게 최신 상품 목록으로 보여주게 할 것인가
- 어떻게 상품리스트 뷰컨트롤러에게 알려줄 것인가

상품리스트 뷰 컨트롤러 스스로가 최신 상품리스트에 대한 정보를 서버에 요청해서, 자신의 테이블뷰를 다시 그리는 것이 
객체의 자율성이 더 높아진다고 생각해서, 상품리스트 뷰컨트롤러에 "통신을 하고, 통신 받은 리스트로 다시 화면을 띄어주는 기능`handleRefreshControl()`"을 구현했습니다. 

처음엔 1페이지에서 상품 1개만 다운받아서, 테이블뷰의 가장 첫번째 row에 올려놓는 방법을 생각했으나, 상품 등록을 하는 사이 다른 사람들이 추가한 상품이 있으면 그건 업데이트를 못해주는 한계가 있었습니다. 
그래서 상품리스트와 관련한 모든 property를 리셋 시켜주고, 상품리스트페이지요청 API를 다시 호출해서 1페이지의 20개 아이템에 대한 정보를 받아와서 `tableView.reloadData()`를 호출하도록 구현했습니다.

```swift
 @objc private func handleRefreshControl() {
        resetProductListPageInfo()
        let request = ProductsListPageRequest(pageNo: 1, itemsPerPage: 20)
        APIExecutor().execute(request) { (result: Result<ProductsListPage, Error>) in
            switch result {
            case .success(let productsListPage):
                self.currentPageNo = productsListPage.pageNo
                self.hasNextPage = productsListPage.hasNext
                self.products.append(contentsOf: productsListPage.pages)
                DispatchQueue.main.async {
                    self.tableView.reloadData()
                    if self.refreshControl?.isRefreshing == false {
                        self.scrollToTop(animated: false)
                    }
                    self.refreshControl?.endRefreshing()
                }
            case .failure(let error):
                // Alert 넣기
                print("ProductsListPage 통신 중 에러 발생 : \(error)")
                return
            }
        }
    }
```
그리고 `handleRefreshControl()`를 호출시키기 위해 다음과 같은 시도를 해보았습니다.

### 1차 시도
`handleRefreshControl()`를 상품등록 excute의 completionHandler에서 직접 호출해주고자 했습니다.  그런데 아래와 같이  `self.presentingViewController`를 `ProductTableViewController`로 타입캐스팅에 실패해서 구현하지 못했습니다.
![Untitled](https://user-images.githubusercontent.com/81469717/150973234-5ab9b9f9-f040-4e80-87fe-fb571d54b521.png)

### 2차 시도
`handleRefreshControl()`를 상품리스트 뷰컨트롤러의 `viewWillAppear()`에서 호출해주고, 
상품등록 뷰컨트롤러 상품등록 execute의 completionHandler에서 `self.presentingViewController`로 상품리스트 뷰컨트롤러의 `viewWillAppear()`를 호출하고자 했습니다. 
그런데 이렇게 하면 상품등록화면에서 상품리스트화면으로의 전환 뿐만 아니라, LIST 화면과 GRID화면의 전환 사이에도 `viewWillAppear()`r가 호출되어 탭 바컨트롤러로 구현한 장점(뷰컨간에 이동해도 마지막 시점을 보여주는 장점)이 사라지는건 아닌가 싶었습니다ㅠㅠ 
또 무거운 작업인`reloadData()`가 호출되는게 부담스럽고, 메모리 낭비가 아닌가 싶기도 했습니다

### 3차 시도
결국 `newProductRegistered`라는 노티피케이션을 정의하고, 
excute의 Handler에서 노티피케이션을 포스트하고, 상품리스트 뷰 컨트롤러들에 Observer를 추가, selector로 `handleRefreshControl()` 호출하는 방법으로 구현했습니다.
<br>

## 2️⃣ 스크롤이 자동으로 되도록 구현 어떻게 하지?

1번에서 코드로서 `refresh` 하는 기능을 구현 성공했는데, 상품리스트 화면으로 돌아갔을 때 화면의 가장 윗 부분으로 위치(시선)가 이동하지 않았습니다. `reloadData()`와, `화면에서 몇 번째 셀을 보여줄지`는 연관관계가 없다는 걸 깨달아서. 어떻게 화면을 위로 올려줄지 고민했습니다. 

테이블뷰 컨트롤러와 컬렉션뷰 컨트롤러의 extension에 가장 첫번째 cell또는 item위치로 이동하도록 하는 메서드를 구현했습니다.
```swift
extension UITableViewController {
    
    func scrollToTop(animated: Bool) {
        let topRow = IndexPath(row: .zero, section: .zero)
        self.tableView.scrollToRow(at: topRow, at: .top, animated: animated)
    }
}
```
그리고 이 메서드를 `reloadData()`가 완료되면 동기적으로 호출되도록 구현했습니다.

<br>

## 3️⃣ Keyboard 가 올라갈 때, 어떻게 UI 컴포넌트들이 가려지지 않게 할까? 
`UIResponder.keyboardWillShowNotification`와 `UIResponder.keyboardWillHideNotification`을 받아서 각 selector 메서드에서  `view.frame.origin.y`을 키보드의 높이만큼 ‘다이나믹하게’ 변경해주는 전략을 선택해보았습니다.
그런데 언뜻 잘 작동하는 것처럼 보이지만, 키보드의 높이 만큼, 스크롤 가능한 영역에서 벗어나버리는 부작용이 생겼습니다😭
해당 고민을 이슈로 작성해보았습니다! https://github.com/Jager-yoo/ios-open-market/issues/5 
<br>

# 🤩궁금한 점

## 1️⃣ UIViewController 타입의 extension 을 활용한 showAlert 메서드 재사용

- 저희는 alert 를 띄워주는 메서드를 `UIViewController` 타입의 extension 에 구현해서 재사용했습니다.
- 그러면 UIViewController 뿐만 아니라, 뷰컨을 상속받는 UITableViewController, UICollectionViewController 에서도 메서드만 호출해서 즉시 사용할 수 있는 편리함이 있기 때문입니다.
- 현업에서도 이렇게 `extension`을 활용해서 alert 를 구현해 재사용하시는지 궁금합니다! 또는 더 좋은 방법이 있을까요? 🤔

## 2️⃣ 테이블뷰와 컬렉션뷰 간의 중복되는 코드 재사용 방법

- 테이블뷰와 컬렉션뷰 사이에 공통 코드가 상당히 많습니다.
- 공통 메서드로 통합을 시도했으나, 대부분의 메서드들이 각각의 뷰컨에서 고유한 프로퍼티인 `tableView`, `collectionView` 를 사용하고 있기 때문에, 메서드를 통합하면 `분기 처리`가 필요해지더라구요. 😅
- 와중에 찰리께서 라이브 코드 리뷰 주실 때 `"분기처리를 경계해라."`라는 조언을 해주신 것도 생각이 나서... 선뜻 메서드를 합치기 망설여졌습니다.
- 위 1️⃣ 번에서 이야기한 `alert` 기능은 위에서 언급한 프로퍼티가 필요하지 않기 때문에, 테이블뷰와 컬렉션뷰의 `공통 조상`인 UIViewController 타입의 extension 으로 분리할 수 있었지만, 다른 메서드들은 어려웠습니다.
- 🙋🏻 질문은 -> tableView, collectionView 프로퍼티처럼, `뷰컨 고유의 프로퍼티만 달라지고 로직은 거의 흡사한 메서드들`이 있다면, 찰리는 어떻게 리팩토링을 시도하실 것 같으신가요? 🤔

## 3️⃣ UIImagePickerController 인스턴스 생성 비용에 대한 의문

- Xcode 의 시뮬레이터로 [상품등록] Modal 에 진입할 때, 버튼을 누르고 1초 이상 시간이 소요되는 현상이 있습니다.
- 저희는 `UIImagePickerController` 인스턴스를 만들어주느라 그런 게 아닐까 가설을 세웠는데요.
- 막상 아이폰에 연결해 앱을 실행해보면, 시뮬레이터와 다르게, [상품등록] Modal 이 아주 빠르게 뜨더라구요.
- 🙋🏻‍♂️ UIImagePickerController 인스턴스의 생성 비용이 큰 것인지, 아니면 시뮬레이터에서의 문제인지 궁금합니다!

## 4️⃣ 키보드 관련 이슈 어떻게 해결하는 게 좋을까?
고민한 점 3번의 이슈를 어떤 방향으로 대응하는 것이 **기기에 상관없이 사용자에게 편집중인 컨트롤을 잘 보이도록** 할 수 있을까요?

고려해볼 만한 해결 포인트를 짚어주시면 감사하겠습니다!😀

<br>

## 🤖 아직 구현하지 못한 기능 
- [ ]  textfield와 textView에 대한 `validation`기능
    - [ ]  **상품명 : 3자 이상 100자 이하**
    - [ ]  **상세설명 : 10 ~ 1000자**
- [ ]  이미지 사이즈 `300KB`이하로 줄이는 기능
- [ ]  정방으로 크롭된 이미지가 올라가도록 이미지 리터칭하는 기능
- [x]  TextView에 글자쓰면 자동으로 스크롤 되는 기능
- [x]  키보드 활성화되어도 뒤에 화면 스크롤 가능하도록 리팩터링

Jager-yoo committed:
feat: 스토리보드로 상품등록 Modal 화면 구현

yeahg-dev committed:
fix: 네비게이션 바 위치가 노치 유무에따라 변경되는 이슈 해결

Jager-yoo committed:
feat: 상품등록 화면에 ScrollView, StackView 삽입

yeahg-dev committed:
feat: ImageAddCollectionViewCell 스토리보드로 구현 후 IBOutlet 연결

Jager-yoo committed:
feat: ImagePicker 기능 구현 중(셀 레이아웃 미완료)

yeahg-dev committed:
fix: pagination이 중복적으로 작동하던 이슈 해결 #4

Jager-yoo committed:
refactor: ImagePicker 커스텀 셀 디자인, 오토레이아웃 조정

yeahg-dev committed:
feat: 최대 첨부가능 이미지 개수 알림 alert 구현

- UIViewController Extension으로 alert를 띄우는 메서드 구현해서 UIViewController를 상속하는 ViewController에서 공통적으로 재사용할 수 있도록 함

Jager-yoo committed:
feat: 상품등록 화면에 UI Components 추가하고 오토레이아웃 조정

yeahg-dev committed:
feat: 키보드의 높이만큼 화면이 올라가는 기능 구현 및 각 텍스트필드 키보드 설정

- NotificationCenter를 활용해 키보드 화면 표시 여부에 따라 뷰의 위치 변경시켜주어 키보드에의해 뷰가 가려지는 이슈 해결
- 텍스트 필드의 내용에 따라 키보드 타입 설정
- 텍스트 필드 편집할 때 clear 버튼이 활성화되는 기능 추가

Jager-yoo committed:
feat: 상품등록 기능 구현

yeahg-dev committed:
feat: productTableViewController에 refreshControl 구현

Jager-yoo committed:
feat: 상품등록 성공 시, NotificationCenter 를 통해 화면을 refresh 할 수 있도록 구현

- 유저가 직접 refresh 를 작동시키는 게 아니라, 상품등록이 성공하면 코드를 통해 refresh 하는 것이므로, 화면을 의도적으로 테이블뷰의 최상단으로 끌어올리는 로직을 추가함

yeahg-dev committed:
refactor: UITableViewController Extension으로 scrollToTop 메서드 분리

Jager-yoo committed:
feat: ProductCollectionViewController 에 화면 새로고침 기능 추가

yeahg-dev committed:
feat: 상품등록 실패시 보여줄 alert 구현

Jager-yoo committed:
feat: 키보드 외 영역 터치하면 키보드 내려가는 기능 구현 #5

yeahg-dev committed:
fix: 이미지 추가 셀이 탭 제스처를 받지 못하는 버그 수정

kcharliek commented:
### 고민한 부분에 대해

1. 상품 등록 성공시, 상품 리스트 뷰컨트롤러를 최신 상품리스트로 업데이트하는 방법

Notification을 사용해서 해결해주셨는데요. Notification 을 통한 객체간 통신 방법은 살아있는 모든 객체가 이벤트를 받을 수 있다는 점에서 앱의 단방향 흐름을 깨버립니다. 비슷하게 Singleton 패턴도 사용하는데 유의하셔야하고요. 그런 점에서 저는 Notification을 안씁니다. Notification으
로만 구현할 수 있는 기능이라면 앱의 흐름을 잘못 설계했다는 반증이 됩니다. 

Child가 Parent에게 이벤트를 보내기 위한 방법은 많습니다. Delegate Pattern, Closure, KVO 등. 한번 다시 고민해보시길 바랍니다.

3. Keyboard 가 올라갈 때, 어떻게 UI 컴포넌트들이 가려지지 않게 할까?

Keyboard가 현재 편집중인 컴포넌트를 '가릴 때' 에만 숨김 처리를 하도록 수정해야할 것 같습니다. 그리고 이 부분은 핵심기능이 아니라고 생각해서요. 구글링 해서 Copy & Paste 하셔도 좋다고 생각합니다.

kcharliek commented:
### 궁금한 점에 대해

1. UIViewController 타입의 extension 을 활용한 showAlert 메서드 재사용

저는 그렇게 작성하지 않습니다만, Alert 스타일이 정해져있다면 method로 케이스를 나누어도 괜찮을 것 같아요.

2. 테이블뷰와 컬렉션뷰 간의 중복되는 코드 재사용 방법

공통 코드가 어떻게 생겼는지를 보아야 리팩토링 방향을 잡아볼 수 있을 것 같습니다. 이 부분은 음성으로 이야기 나눠보시죠~

3. UIImagePickerController 인스턴스 생성 비용에 대한 의문

https://stackoverflow.com/questions/8230907/uiviewcontroller-mysteriously-slow-to-load

구글링 해보니 Simulator에서만 느리다고 하네요. 실기기로 빌드 한번 해보시겠어요~?
검색 키워드는 'ios UIImagePickerController performance ' 입니다. 이후에 비슷하게 성능이슈로 고민이 생긴다면 performance라는 키워드를 사용하여 구글링해보세요.

4. 키보드 관련 이슈 어떻게 해결하는 게 좋을까?

위에서 말씀 드린 것 처럼. Keyboard가 현재 편집중인 컴포넌트를 '가릴 때' 에만 숨김 처리를 하도록 수정해야할 것 같습니다.

kcharliek commented on OpenMarket/OpenMarket/Cell/ImagePickerCollectionViewCell.swift:
```diff
@@ -0,0 +1,37 @@
+    }
+    
+    func updateAddedImageCountLabel(images: [UIImage]) {
+        let maximumImageCount = 5
```
5개 까지 노출되어야한다는 건 '비즈니스 로직' 입니다. View에는 View를 구성하는 코드만 있는 것이 좋아요.
View 내부에서 5개 까지만 보여주도록 걸러내는 것 보다, 5개 이상의 images를 생성하지 않고 전달하지 않도록 하는 것이 좋은 방향입니다. 실제로 아래 ImagePicker 띄울 때 5개 제한하는 것 같아서 위 코드 없어도 될 것 같아요.

Jager-yoo replied:
위 `updateAddedImageCountLabel` 메서드는 사진 추가 버튼에 해당하는 Cell 의 `Label`을 꾸며주기 위한 메서드입니다!
(이미지 첨부합니다.)
이 메서드는 `이미지는 5개 까지`만 걸러내는 로직이 없습니다!

상품등록 뷰컨의 `didSelectItemAt` 메서드에 이미지 5개 제한 로직이 걸려있어서, 이미지가 꽉 찼다면 아예 버튼이 눌리지 않게 guard문으로 막아주고 있습니다.


따라서 위 코드는 비즈니스 로직에 의한 분기처리는 없고, UILabel, 즉 `"view"` 를 구성하는 코드만 있다고 생각합니다. 😅 
`let maximumImageCount = 5` 을 분리한 이유는 하드코딩을 방지하고 다른 개발자가 봤을 때 가독성을 높이기 위해서였습니다.

<p align="left"><img src="https://user-images.githubusercontent.com/71127966/151330548-95d6e940-1d63-4e4f-9d8b-fddff4ad1a27.png" width="30%"></p>

- 현재 아쉬운 점은 `maximumImageCount` 라는 변수가 여러 파일에서 공통적으로 쓰이지 못하고 있다는 점입니다.
- 만약 비즈니스 로직이 변경돼서 최대 이미지가 6장으로 변경되는 상황을 가정한다면, `네임스페이스`를 통해 비즈니스 로직을 관리해주는 게 좋겠네요!

kcharliek replied:
아 제가 잘못 이해했네요. 말씀 주신 작성의도 이제 이해했습니다 ㅎㅎ
maximumImageCount를 한 곳에서 관리하지 못하는 점에 대해 아쉬운 점 말씀해주셨고, 네임스페이스 사용해서 해결하는 방향까지 잘 고민해주셨습니다. 이견 전혀 없습니다. 고민 잘해주셨어요!!!

kcharliek commented on OpenMarket/OpenMarket/Controller/ProductRegistrationViewController.swift:
```diff
@@ -0,0 +1,214 @@
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        
+        guard let productImageCollectionView = productImageCollectionView else {
+            return UICollectionViewCell()
```
UICollectionViewCell() 은 크래시를 유발합니다. 사용하지 말아주세요. 

productImageCollectionView 는 delegate method의 파라미터로 전달되니 이것을 사용해주세요.

Jager-yoo replied:
해당 메서드의 파라미터로 들어오는 `collectionView`로 대체해주었습니다!
덕분에 불필요한 guard let 을 한 번 줄이고 로직도 심플해졌습니다. 😄 예리하게 봐주셔서 감사합니다 !!

kcharliek commented on OpenMarket/OpenMarket/Controller/ProductTableViewController.swift:
```diff
@@ -91,4 +95,48 @@ final class ProductTableViewController: UITableViewController {
     }
+    
+    private func addNotificationObserver() {
+        NotificationCenter.default.addObserver(
```
위에서 코멘트 드렸습니다. Notification 은 불가항력일 때에만 사용해주세요.

yeahg-dev replied:
노티피케이션을 델리게이트 패턴으로 대체해서 구현해보았습니다!

1. `RefreshDelegate` 라는 프로토콜을 정의하고 `refresh()` 메서드를 정의했습니다.
2. `ProductTableViewController`와 `ProductCollectionViewController`가 위 프로토콜을 채택하도록 했습니다.
3. `ProductRegisterationViewController`가 `RefreshDelegate`타입의 프로퍼티를 갖게하고, 노티를 포스트 하던 자리에서 `refreshDelegate.refresh()`로 일을 시키도록 했습니다.
4. `ProductRegisterationViewController`의 delegate를 지정하는 로직은 `TabBarViewController`의 `prepare(for:)에서 구현했습니다.

해당 커밋 https://github.com/Jager-yoo/ios-open-market/commit/65e13c83ad7b26244c9e4fcfbe36d7074db95f5d 첨부합니다😊

그런데 얘기하신 방향대로 잘 구현한건지 궁금합니다🙄

Jager-yoo replied:
### ⛑️ 강한 순환 참조 방지를 위한 `weak` 키워드 추가 선언

1. `RefreshDelegate` 프로토콜이 `AnyObject` 프로토콜을 채택하게 만들어서 오직 `클래스` 타입만이 해당 프로토콜을 채택할 수 있게 만들었습니다.
2. delegate 프로퍼티에 `weak` 키워드를 추가해줬습니다.

kcharliek replied:
YES!! Delegate Pattern 작성 아~ 주 잘해주셨습니다.
Class Type만 Reference를 가질 수 있죠. 그래서 Delegate Pattern은 Class만 따를 수 있습니다.
weak로 순환참조 방지까지 아주 잘했어요. !!

Jager-yoo replied:
### 🩹 추가 리팩토링 사항

NotificationCenter 로 구현했을 때는, 상품 등록에 성공하면 2개 뷰컨(테이블/컬렉션)에 신호를 보내서 각각 `리프레시`를 해주고 있었습니다! -> 그래서 사용자가 상품 등록에 성공하면 UI 에 그 상품이 바로 보이도록 해주었는데요!

Delegate Pattern 으로 변경한 뒤에는, `상품등록 모달`을 띄우는 시점에서, 사용자가 쳐다보고 있던 뷰컨(테이블 or 컬렉션) 하나만 리프레시되고, 다른 뷰컨은 리프레시되지 않더라구요!

어떻게 수정해볼까 고민하다가
delegate 프로퍼티를 2개로 만들어서, 그 각각에 일을 시키는 구조로 바꿨습니다! 😄 

```swift
self.showAlert(title: "상품이 성공적으로 등록됐습니다", message: "🤑") { _ in
    self.dismiss(animated: true) {
        self.tableViewRefreshDelegate?.refresh()
        self.collectionViewRefreshDelegate?.refresh()
    }
}
```

kcharliek commented on OpenMarket/OpenMarket/Controller/ProductRegistrationViewController.swift:
```diff
@@ -0,0 +1,214 @@
+            return UICollectionViewCell()
+        }
+        
+        if indexPath.item == .zero {
```
IndexPath를 사용해서 Cell Type을 분기하는 것은 좋지않습니다.
이유는 비즈니스로직이 View 코드와 합쳐져있기 때문인데요.
위 코드라면 만일 Piker Cell의 indePath를 바꾸고싶다면 View코드를 수정해야해요.
'로직이 바뀌면 코드가 바뀌어야하는 건 당연한거 아니야?' 라고 생각하실 수 있겠지만 비즈니스 로직 수정은 Model Layer에서 이루어져야 좋습니다. 정확히 '왜 좋은가' 에 대해, 그러면 어떻게 구현해야 하는가에 대해서는 말이 길어질 것 같아서 음성으로 설명드리겠습니다.

yeahg-dev replied:
indexPath에 따라 분기 처리해주는 로직을 없애기 위해 CellType 배열을 정의해주었습니다.
이 배열은 indexPath.item에 따라 보여질 CellType들을 가지는 배열입니다. 따라서 처음엔 `.pickerCell`을 가지고 있습니다.
```swift
    private var cells: [CellType] = [.pickerCell]
```
그리고 PickerController로 `productImages`에 이미지가 추가될 때마다 `cells.append(.imageCell)`로 cellType을 추가해줬습니다. 

그리고 `cellForItemAt`에서 `indexPath.item`으로 `cells`의 타입을 가져와서 각각 CellType을 만들어서 반환하도록 해주었습니다. 그런데 이렇게 해도 결국 indexPath에 의존하고 있고, 비지니스로직에 의해 결정되는게 아닌 것 같아서ㅠㅠ 찰리가 조언주신 방향이 아닌 것 같습니다...😢 

```swift
enum CellType {
        case pickerCell
        case imageCell
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        // 결국 indexPath에 따라 반환할 Cell을 정해주는 로직이 필요...
        // 그 로직을 Model로 가져가자
        // ProductImages가 0일때 -> pickerCell만 반환
        // ProductImages가 1이상 일 때 -> pickerCell, imageCell을 반환 🆘 여기서 문제. 결국 indexPath에 따라 Cell을 분기처리해야함.
        // 그래서 [CellType] 배열을 만듦

        let cellType: CellType = cells[indexPath.item]
        
        switch cellType {
        case .imageCell:
            let cell = collectionView.dequeueReusableCell(withClass: ProductImageCollectionViewCell.self, for: indexPath)
            let targetImage = productImages[safe: indexPath.item - 1]
            cell.updateProductImageView(image: targetImage)
            return cell
        case .pickerCell:
            let cell = collectionView.dequeueReusableCell(withClass: ImagePickerCollectionViewCell.self, for: indexPath)
            cell.updateAddedImageCountLabel(images: productImages)
            return cell
        }
    }
```

아무리  생각해도  길이 보이지 않스습니다...
도와주세요 찰리🆘

kcharliek replied:
와우!! 제가 조언드린 방향이 맞습니다.
indexPath 를 사용한 코드기 때문에 틀렸다고 생각하셨을까요?
제 코멘트의 핵심은 indexPath를 이용해서 View를 그리지 않기를 바랬던 것이지, indexPath로 cellType을 가져오는 것까지 막으려고 했던 것은 아닙니다. 작성해주신 코드를 보면 cells와 productImages를 조작하는 것 만으로 View가 어떻게 그려질지를 결정할 수 있습니다. 비즈니스 로직을 설계하고 model을 조작하는 것은 어쩌면 당연해요. 기존 코드 처럼 '0번 째 Cell은 pickerCell 이다' 라고 박아둔다면, 로직 변경을 위해 View코드를 건드릴 수 밖에 없겠죠. '비즈니스 로직 수정을 위해 Model' 을 수정할 것인가, '비즈니스 로직 수정을 위해 View를 수정할 것인가' 에서 우리는 Model을 수정하기로 했고, 그로 인해 유지보수가 용이해지고, 자유도가 높아지고, 가독성 또한 좋아졌습니다. 완전히 위 맥락을 이해하지 못하셔도 괜챃아요. 'View는 Model을 담는 껍데기 일 뿐이다', 'Model이 View를 결정한다' 라는 관점만 얻어가신다면 많이 배우셨다고 생각합니다 :)

yeahg-dev replied:
아하~~! 찰리가 말씀하신 목적이 조금은 그려지는 것 같아요!
**View는 Model을 담는 껍데기 일 뿐이다**, **Model이 View를 결정한다** 이 관점! 다시 한번 뇌에 새기구 앞으로 설계할 때 고민해보도록하겠습니다😊

kcharliek commented on OpenMarket/OpenMarket/Controller/ProductRegistrationViewController.swift:
```diff
@@ -0,0 +1,214 @@
+        self.dismiss(animated: true)
+    }
+    
+    @IBAction private func doneButtonTapped(_ sender: UIBarButtonItem) {
```
doneButtonTapped 내부 구성이 복잡해보입니다.
단일책임원칙을 생각하며 로직을 쪼개주세요.

Jager-yoo replied:
`doneButtonTapped` 메서드 내부에서 해주는 작업을 요약하자면 `상품등록 리퀘스트를 만들어서 보내고 결과 처리`라고 볼 수 있습니다.
그래서 아래와 같은 메서드로 분리시켰습니다.

```swift
@IBAction private func doneButtonTapped(_ sender: UIBarButtonItem) {
    handleProductRegistrationRequest()
}
```

그렇게 분리시킨 메서드에서의 작업을 크게 3개로 나눠서 다시 분리시켰습니다.

1. 새롭게 등록하는 상품의 인스턴스를 만드는 작업 -> `createNewProductInfo()` 메서드로 분리
2. 새롭게 등록하는 상품의 이미지들의 인스턴스를 만드는 작업 -> `createImageFiles(newProductName:)` 메서드로 분리
3. request 를 만들고 통신하는 작업 -> 분리 안 함

Jager-yoo committed:
fix: 키보드 올라오면 스크롤 가능한 영역이 줄어드는 이슈 해결 #5

- 기존에는 view.frame.origin.y 에 키보드 높이를 반영해줬다면, 이번 커밋에서는 scrollView.contentInset.bottom 에 키보드 높이 만큼 값을 넣어줌
- 스토리보드에서 scrollView 의 keyboard 설정 중 'Dismiss on drag' 를 켜서, 드래그를 시작하면 키보드가 자동으로 내려가는 기능 추가함

yeahg-dev committed:
refactor: 상품등록 성공시 화면 새로고침 유발 로직을 NotificationCenter에서 Delegate pattern으로 리팩터링

Jager-yoo committed:
refactor: productImageCollectionView 의 cellForItemAt 메서드 내부 불필요한 로직 삭제

yeahg-dev committed:
refactor: 강한 순환 참조를 방지하기위해 delegate 프로퍼티에 weak 키워드 추가

Jager-yoo committed:
refactor: doneButtonTapped 메서드 내부 로직 중, 사용자 입력값 받고 NewProductInfo 반환하는 부분을 별도 메서드로 분리

- 불필요한 UITextFieldDelegate 익스텐션 삭제하고 메서드 위치 변경

yeahg-dev committed:
refactor: doneButtonTapped 메서드 내부 로직 중 NewProductInfo의 ImageFile 배열 반환하는 부분을 별도 메서드로 분리

kcharliek reviewed (approved):
고생많았어요. 아주 잘해주셨습니다.

kcharliek commented on OpenMarket/OpenMarket/Controller/ProductCollectionViewController.swift:
```diff
@@ -108,4 +109,48 @@ final class ProductCollectinoViewController: UICollectionViewController {
+    @objc private func handleRefreshControl() {
+        resetProductListPageInfo()
+        let request = ProductsListPageRequest(pageNo: 1, itemsPerPage: 20)
+        APIExecutor().execute(request) { (result: Result<ProductsListPage, Error>) in
```
completion block(escaping closure) 에서 self를 strong 참조하셨습니다.
API결과가 영원히 응답이 없다면 self가 return될 일은 없을 것 같은데요. 어떻게 참조해야할지 '구글링' 해보세요. 
키워드는 'ios escaping closure weak self' 입니다 :)

yeahg-dev replied:
`[weak self]` 로 참조해야 강한 순환 참조를 방지할 수 있을 것 같은데요~

먼저 저희가 이해한 바를 설명드리자면
우선 클로저도 레퍼런스타입이고, 탈출클로저는 만들어지면서 값을 캡쳐하며 관련 값을 저장할 메모리 공간을 만들게 됩니다.
`handleRefreshControl()`이 호출되면서 `excute` 메서드의 탈출클로저는 `self`를 캡처합니다. 캡처함과 동시에 `self`(ProductCollecitionViewController)의 RC를 1 늘립니다. 
그리고 `self`도 클로저의 RC를 1을 늘립니다. (이로서 둘 사이에 강한 순환참조가 발생함)

클로저가 deinit될 때 캡처한 `self`의 RC도 -1을 해줍니다.
클로저가 deinit되는 시점이라 하면 클로저가 실행된 직후입니다. = 즉, 클로저가 실행완료되면 `self`가 올려주었던 RC는 -1이 된다고 이해했습니다.(혹은 클로저를 어떠한 변수에서 저장하고 있다면 변수 = nil이되고, RC가 0인 상태) 
그런데 `excute`의 클로저 같은 경우 response가 오지 않는다면 영원히 실행되지 않을 것이기에,
`self`의 RC도 여전히 1을 늘려준채로 머무릅니다.

이와 같은 문제상황을 해결하기 위해 `self`를 `weak`로 참조하여 RC를 늘려주지 않게 대응한다. 라고 이해했습니다. 

찰리가 말씀하신 "API결과가 영원히 응답이 없다면 self가 return될 일은 없을 것 같은데요"에서
`self가 return된다` = `self의 RC를 -1해준다`라고 이해했는데 맞게 이해한게 맞을까요???🙄

kcharliek replied:
제 표현이 조금 구린 것 같아요. 'self가 return된다' = 'self의 RC를 -1해준다' 로 이해 잘 해주셨습니다.

위쪽에 원인파악해주신 부분들 모두 '정확히' 맞습니다. 조사 잘 해주셨습니다. 이후에 escaping closure를 사용하게된다면 이 부분 신경 써주세요~

Jager-yoo committed:
refactor: doneButtonTapped 메서드 내부 로직을 별도 메서드인 handleProductRegistrationReqeust() 로 분리하여 단일책임원칙 준수

yeahg-dev committed:
file: ImageCacheManager 파일 Utility폴더로 이동

yeahg-dev committed:
refactor: 뷰와 비지니스로직 분리를 위해 CellType을 정의해서 리팩터링

Jager-yoo committed:
refactor: 상품등록 화면의 스택뷰 내부 컴포넌트들에 명시적인 height 부여

- 스택뷰를 UI 컴포넌트들로 채울 때, Intrinsic Content Size 를 사용하기 보다는 정확한 높이 Constant 를 부여했음

yeahg-dev committed:
fix: 상품등록 성공시 테이블뷰, 컬렉션뷰 모두 refresh되도록 수정

- tableViewRefreshDelegate, collectionViewRefreshDelegate를 각각 만들어주고, TabBarViewController에서 각각 delegate를 지정하도록 함

Jager-yoo committed:
refactor: escaping 클로저에서 self 를 강한 참조하지 않도록 [weak self] 키워드 삽입하여 리팩토링

yeahg-dev committed:
style: 줄바꿈 수정
